src/node_sqlite.cc

```diff
@@ -773,15 +773,20 @@ void DatabaseSync::ApplyChangeset(const FunctionCallbackInfo<Value>& args) {
         options->Get(env->context(), env->onconflict_string()).ToLocalChecked();
 
     if (!conflictValue->IsUndefined()) {
-      if (!conflictValue->IsNumber()) {
+      if (!conflictValue->IsFunction()) {
         THROW_ERR_INVALID_ARG_TYPE(
             env->isolate(),
-            "The \"options.onConflict\" argument must be a number.");
+            "The \"options.onConflict\" argument must be a function.");
         return;
       }
-
-      int conflictInt = conflictValue->Int32Value(env->context()).FromJust();
-      conflictCallback = [conflictInt]() -> int { return conflictInt; };
+      Local<Function> conflictFunc = conflictValue.As<Function>();
+      conflictCallback = [env, conflictFunc](int conflictType) -> int {
+        Local<Value> argv[] = {Integer::New(env->isolate(), conflictType)};
+        Local<Value> result =
+            conflictFunc->Call(env->context(), Null(env->isolate()), 1, argv)
+                .ToLocalChecked();
+        return result->Int32Value(env->context()).FromJust();
+      };
```

Comment:
Similar to passing `{ onConflict: SQLITE_CHANGESET_REPLACE }` previously, this will still throw if the user returns an illegal resolution code:

```js
db.applyChangeset(changeset, {
  // let's assume this returns SQLITE_CHANGESET_REPLACE illegally,
  // or returns something that isn't a resolution code
  onConflict: (eConflict) => SOME_ILLEGAL_RETURN_VALUE
})
// Uncaught [Error: not an error] {
//   code: 'ERR_SQLITE_ERROR',
//   errcode: 0,
//   errstr: 'not an error'
// }
```

The way I see it, there are two ways of dealing with this eventuality:
- Wrap the call to the user callback in a function that validates its output, testing that the value returned from the callback is one of the sqlite conflict resolution codes, and that `SQLITE_CHANGESET_REPLACE` isn't returned illegally.
  - If validation fails, then return `SQLITE_CHANGESET_ABORT` to sqlite, allow the library to clean up gracefully, then throw a descriptive runtime error.
- Don't validate the output from the user callback, and allow `sqlite3changeset_apply()` to fail, rollback, and return `SQLITE_MISUSE` if the value returned from the callback is invalid.
  - If this is the case, then the error handling in node-sqlite needs to be improved to ensure that the thrown error makes sense, for the reasons mentioned in #56210.

Reply:
Good point.

Reply:
Done. I just read the message with `sqlite3_errstr`.